### ISSUE TITLE
Graceful shutdown unresponsive to shutdownGraceMs setting

### ISSUE THREAD

**Issue by jacob-go** [linkerd/linkerd#1887]
## Filing a Linkerd issue ##

Issue Type:

- [X] Bug report

**What happened**:
I was testing linker-to-linker architecture. Sent a sequence of non-stop requests to an endpoint behind a Linkerd proxy and restarted the proxy without stoping the requests. I witnessed that no matter the load, Linkerd get killed in about 2 seconds. 
I've searched after a setting for this period of time before termination, maybe explicitly configuring a larger interval. I've configured my `admin` block in the configuration file with `shutdownGraceMs: 10000` but still - Linkerd is terminated after 2 seconds. 

**What you expected to happen**:
A) A more cohesive graceful shutdown, closing sockets/connections, draining the queue of requests until none is there and then terminating. 
B) Linkerd task would be terminated after the period of time configured in the `admin` block. 

**How to reproduce it (as minimally and precisely as possible)**:
```
---
admin:
  ip: 0.0.0.0
  port: 4447
  shutdownGraceMs: 10000
```

Then see exact time of SIGTERM and exit in `stderr`:
```
I 0329 08:14:48.002 UTC THREAD29 TraceId:b229fc3755c0e29f: [/admin/shutdown] from <TASK_IP> quitting
I0329 08:14:49.415153 18340 health_checker.cpp:165] Health checking stopped
W0329 08:14:49.415153 18337 logging.cpp:91] RAW: Received signal SIGTERM from process 5704 of user 0; exiting
```

**Anything else we need to know?**:
I think that maybe [this](https://github.com/linkerd/linkerd/issues/1834) issue is caused by the graceful shutdown issue. 

**Environment**:
- linkerd/namerd version, config files: the only relevant config is in the `admin` block
- Platform, version, and config files (Kubernetes, DC/OS, etc): DC/OS
- Cloud provider or hardware configuration: AWS

**Comment by jacob-go** [comment]
Hi there @dadjeibaah, how are you? 

Did you've managed to take a look? Do you see a similar behaviour?

**Comment by dadjeibaah** [comment]
Hi @jacob-go, thanks for checking in! I have not had the chance to look into this issue yet. However, I was able to reproduce the issue pretty easily based on your description. We will update the issue once we come around to a possible fix for this issue.

**Comment by dadjeibaah** [comment]
I've taken a look at the shutdown grace period issue and this is what I understand so far based on the three tests I run:

- **Test 1**: If no requests are in flight and linkerd gets a SIGTERM, it will shutdown immediately even if `shutdownGraceMs` is set.
- **Test 2**: If a request is in flight and is completed before the graceful shutdown period completes, linkerd will shutdown immediately after the completed request. Meanwhile, no requests will be processed during the graceful shutdown period.
- **Test 3**: If the request is never completed within the graceful shutdown window, Linkerd will force shutdown and drop the in-flight request. Just like the scenario above, any request that comes in during the graceful period shutdown will not be processed.

### No in-flight requests with 10 second graceful shutdown window

```bash
I 0412 13:57:48.114 PDT THREAD1: initialized
D 0412 13:57:48.498 PDT THREAD20:
I 0412 13:58:14.322 PDT THREAD23: Received SIGTERM. Shutting down ...
```
### Request with a 8 second latency and linkerd with a 10 second graceful shutdown window
```bash
I 0412 13:44:03.681 PDT THREAD1: initialized
D 0412 13:44:04.069 PDT THREAD22:
D 0412 13:44:57.328 PDT THREAD24 TraceId:883e1c5975bad4db: LoadService: loaded instance of class com.twitter.server.FlagResolver for requested service com.twitter.finagle.Resolver
D 0412 13:44:57.333 PDT THREAD24 TraceId:883e1c5975bad4db: LoadService: loaded instance of class com.twitter.finagle.zookeeper.ZkResolver for requested service com.twitter.finagle.Resolver
D 0412 13:44:57.352 PDT THREAD24 TraceId:883e1c5975bad4db: LoadService: loaded instance of class com.twitter.finagle.serverset2.Zk2Resolver for requested service com.twitter.finagle.Resolver
I 0412 13:44:57.366 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[inet] = com.twitter.finagle.InetResolver(com.twitter.finagle.InetResolver@28428c1f)
I 0412 13:44:57.366 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[fixedinet] = com.twitter.finagle.FixedInetResolver(com.twitter.finagle.FixedInetResolver@4982f0e1)
I 0412 13:44:57.367 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[neg] = com.twitter.finagle.NegResolver$(com.twitter.finagle.NegResolver$@76802a73)
I 0412 13:44:57.367 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[nil] = com.twitter.finagle.NilResolver$(com.twitter.finagle.NilResolver$@4a1e72c8)
I 0412 13:44:57.367 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[fail] = com.twitter.finagle.FailResolver$(com.twitter.finagle.FailResolver$@14459692)
I 0412 13:44:57.367 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[flag] = com.twitter.server.FlagResolver(com.twitter.server.FlagResolver@6456b4db)
I 0412 13:44:57.367 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[zk] = com.twitter.finagle.zookeeper.ZkResolver(com.twitter.finagle.zookeeper.ZkResolver@13f2c990)
I 0412 13:44:57.368 PDT THREAD24 TraceId:883e1c5975bad4db: Resolver[zk2] = com.twitter.finagle.serverset2.Zk2Resolver(com.twitter.finagle.serverset2.Zk2Resolver@2a436114)
I 0412 13:44:59.028 PDT THREAD29: Received SIGTERM. Shutting down ...
I 0412 13:45:05.586 PDT THREAD24: Reaping /svc/dog
I 0412 13:45:05.589 PDT THREAD24: Reaping $/inet/127.1/7777
```

### Request with a 15 second latency with linkerd shutdown grace period of 10 seconds
```bash
I 0412 13:51:29.506 PDT THREAD1: initialized
D 0412 13:52:05.671 PDT THREAD28 TraceId:9c28ecd4a090f466: LoadService: loaded instance of class com.twitter.server.FlagResolver for requested service com.twitter.finagle.Resolver
D 0412 13:52:05.676 PDT THREAD28 TraceId:9c28ecd4a090f466: LoadService: loaded instance of class com.twitter.finagle.zookeeper.ZkResolver for requested service com.twitter.finagle.Resolver
D 0412 13:52:05.693 PDT THREAD28 TraceId:9c28ecd4a090f466: LoadService: loaded instance of class com.twitter.finagle.serverset2.Zk2Resolver for requested service com.twitter.finagle.Resolver
I 0412 13:52:05.709 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[inet] = com.twitter.finagle.InetResolver(com.twitter.finagle.InetResolver@3694dcec)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[fixedinet] = com.twitter.finagle.FixedInetResolver(com.twitter.finagle.FixedInetResolver@6bb562c6)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[neg] = com.twitter.finagle.NegResolver$(com.twitter.finagle.NegResolver$@3e4bcd6f)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[nil] = com.twitter.finagle.NilResolver$(com.twitter.finagle.NilResolver$@7cc3af3a)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[fail] = com.twitter.finagle.FailResolver$(com.twitter.finagle.FailResolver$@6c206ff2)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[flag] = com.twitter.server.FlagResolver(com.twitter.server.FlagResolver@2cc383e2)
I 0412 13:52:05.710 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[zk] = com.twitter.finagle.zookeeper.ZkResolver(com.twitter.finagle.zookeeper.ZkResolver@4c8303a0)
I 0412 13:52:05.711 PDT THREAD28 TraceId:9c28ecd4a090f466: Resolver[zk2] = com.twitter.finagle.serverset2.Zk2Resolver(com.twitter.finagle.serverset2.Zk2Resolver@2d58df75)
I 0412 13:52:07.346 PDT THREAD32: Received SIGTERM. Shutting down ...
D 0412 13:52:17.357 PDT THREAD31 TraceId:9c28ecd4a090f466: Failed mid-stream. Terminating stream, closing connection
com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /127.0.0.1:7777. Remote Info: Not Available
	at com.twitter.finagle.netty4.transport.ChannelTransport$$anon$1.channelInactive(ChannelTransport.scala:188)
...

D 0412 13:52:17.375 PDT THREAD28 TraceId:9c28ecd4a090f466: Exception propagated to the default monitor (upstream address: /127.0.0.1:56577, downstream address: /127.0.0.1:7777, label: $/inet/127.1/7777).
com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /127.0.0.1:7777, Downstream label: $/inet/127.1/7777, Trace Id: 9c28ecd4a090f466.c31ef3bf3fb5c7fb<:9c28ecd4a090f466

I 0412 13:52:17.473 PDT THREAD28: Reaping /svc/dog
I 0412 13:52:17.478 PDT THREAD28: Reaping $/inet/127.1/7777
E 0412 13:52:17.481 PDT THREAD28 TraceId:9c28ecd4a090f466: service failure: com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /127.0.0.1:7777, Downstream label: $/inet/127.1/7777, Trace Id: 9c28ecd4a090f466.c31ef3bf3fb5c7fb<:9c28ecd4a090f466
```

**Comment by jacob-go** [comment]
Hello, @dadjeibaah, thanks for investigating. Unfortunately, we see something different. 

## No in-flight requests with 10 second graceful shutdown window
```
I 0423 13:23:14.725 UTC THREAD36 TraceId:390ad22c14dedea3: [/admin/shutdown] from 127.0.0.1 quitting
I0423 13:23:16.155283 18896 health_checker.cpp:165] Health checking stopped
W0423 13:23:16.155283 18889 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
``` 

## Request with a 8 second latency and linkerd with a 10 second graceful shutdown window
```
I 0423 13:25:10.461 UTC THREAD38 TraceId:3c02c1ff15ea5a99: [/admin/shutdown] from 127.0.0.1 quitting
I 0423 13:25:11.536 UTC THREAD33 TraceId:3c02c1ff15ea5a99: Reaping /svc/check-linkerd-deployments.dcos-internal-test.mars-fra-apps.zooz.co
I 0423 13:25:11.539 UTC THREAD33 TraceId:3c02c1ff15ea5a99: Reaping #/io.l5d.marathon/check-linkerd-deployments
E 0423 13:25:11.541 UTC THREAD33 TraceId:c1050292e20dffdf: service failure: com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /10.1.3.250:7606, Downstream label: #/io.l5d.marathon/check-linkerd-deployments, Trace Id: c1050292e20dffdf.bd83b8fbbae6750e<:c1050292e20dffdf
I0423 13:25:12.947687 20386 health_checker.cpp:165] Health checking stopped
W0423 13:25:12.947687 20382 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```

## Request with a 15 second latency with linkerd shutdown grace period of 10 seconds
```
I 0423 13:27:10.218 UTC THREAD38 TraceId:ae1cffdecca59d9d: [/admin/shutdown] from 127.0.0.1 quitting
I 0423 13:27:11.278 UTC THREAD23 TraceId:ae1cffdecca59d9d: Reaping /svc/check-linkerd-deployments.dcos-internal-test.mars-fra-apps.zooz.co
I 0423 13:27:11.282 UTC THREAD23 TraceId:ae1cffdecca59d9d: Reaping #/io.l5d.marathon/check-linkerd-deployments
E 0423 13:27:11.284 UTC THREAD23 TraceId:5901fbfd51cee73d: service failure: com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /10.1.3.250:7606, Downstream label: #/io.l5d.marathon/check-linkerd-deployments, Trace Id: 5901fbfd51cee73d.95057738731a166e<:5901fbfd51cee73d
I0423 13:27:12.717963 24235 health_checker.cpp:165] Health checking stopped
W0423 13:27:12.717963 24233 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```

I wondered wether it's because the version, I'm working on Linekrd 1.3.0. But Linkerd 1.3.7 shows the same results:

## No in-flight requests with 10 second graceful shutdown window
```
I 0423 13:29:52.533 UTC THREAD1: initialized
I 0423 13:30:20.145 UTC THREAD37 TraceId:2300803fdb35abf6: [/admin/shutdown] from 127.0.0.1 quitting
I0423 13:30:21.590337 28055 health_checker.cpp:165] Health checking stopped
W0423 13:30:21.590337 28045 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```

## Request with a 8 second latency and linkerd with a 10 second graceful shutdown window
```
I 0423 13:30:32.854 UTC THREAD1: initialized
I 0423 13:31:15.368 UTC THREAD29 TraceId:a325ba805411f1e6: [/admin/shutdown] from 127.0.0.1 quitting
I 0423 13:31:16.434 UTC THREAD24 TraceId:a325ba805411f1e6: Reaping /svc/check-linkerd-deployments.dcos-internal-test.mars-fra-apps.zooz.co
I 0423 13:31:16.437 UTC THREAD24 TraceId:a325ba805411f1e6: Reaping #/io.l5d.marathon/check-linkerd-deployments
E 0423 13:31:16.440 UTC THREAD24 TraceId:0d4cf1d5f9aca468: service failure: com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /10.1.3.250:7606, Downstream label: #/io.l5d.marathon/check-linkerd-deployments, Trace Id: 0d4cf1d5f9aca468.c5aae1fb78ec8dd0<:0d4cf1d5f9aca468
I0423 13:31:17.870793 28864 health_checker.cpp:165] Health checking stopped
W0423 13:31:17.870793 28860 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```

## Request with a 15 second latency with linkerd shutdown grace period of 10 seconds
```
I 0423 13:32:05.127 UTC THREAD1: initialized
I 0423 13:32:31.389 UTC THREAD39 TraceId:ed7cd5fee810c8ef: [/admin/shutdown] from 127.0.0.1 quitting
I 0423 13:32:32.456 UTC THREAD23 TraceId:ed7cd5fee810c8ef: Reaping /svc/check-linkerd-deployments.dcos-internal-test.mars-fra-apps.zooz.co
I 0423 13:32:32.460 UTC THREAD23 TraceId:ed7cd5fee810c8ef: Reaping #/io.l5d.marathon/check-linkerd-deployments
E 0423 13:32:32.462 UTC THREAD23 TraceId:aa0dd291f1db2156: service failure: com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: Not Available, Upstream id: Not Available, Downstream Address: /10.1.3.250:7606, Downstream label: #/io.l5d.marathon/check-linkerd-deployments, Trace Id: aa0dd291f1db2156.4a60e844d1900f81<:aa0dd291f1db2156
I0423 13:32:33.847641 31006 health_checker.cpp:165] Health checking stopped
W0423 13:32:33.847641 31006 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```



We're working on top of DC/OS 1.9.6, but the SIGTEM is sent from local machine via `http://127.0.0.1/admin/shutdown` endpoint. 

Can you elaborate more on your experiments?

**Comment by dadjeibaah** [comment]
It is interesting to see the linkerd process consistently shuts down within 2 seconds.  My experiments were conducted locally. I had a dummy server that accepted a number of requests that responded with some preconfigured latency i.e. one request with an 8 second latency and another with a 15 second latency.  For my first test, I simply sent a `SIGTERM` to linkerd with no in-flight requests . I then ran two tests where I started linkerd, issued a request through linkerd to the dummy server, initiated a `SIGTERM` on the linkerd process and watched the logs to observe how long it took for the process to terminate. I did not use the shutdown endpoint but that might be something I could possibly introduce in my tests. 

I am a little curious about this line in the logs you provided:
```bash
W0423 13:32:33.847641 31006 logging.cpp:91] RAW: Received signal SIGTERM from process 3921 of user 0; exiting
```
I don't recall seeing that come up in the linkerd logs. I am not entirely sure of how DC/OS works but is that a process that runs along side linkerd? If so, does that process interact with Linkerd in anyway?

**Comment by jacob-go** [comment]
`3921` is the mesos-agent process on that instance:
![architecture-example](https://user-images.githubusercontent.com/29705346/39142735-21273662-4734-11e8-9529-ba57e9c5b193.jpg)

I'll try to send a simple SIGTERM directly to the container, but why do you think this influences the results? Should a SIGTERM behave differently than `/admin/shutdown`, which in turn behaves differently than sending a restart command via Marathon API?

**Comment by dadjeibaah** [comment]
I don't think, OTTOH, that sending a `SIGTERM` vs using the `/admin/shutdown` are any different.  I would like to make sure the tests closely match how the issue is manifested just in case the `/admin/shutdown` has some additional tear down tasks that I am not aware of.

**Comment by jacob-go** [comment]
Please let me know if I can help with something

**Comment by dadjeibaah** [comment]
Thanks, I am going to rerun my tests but this time with the admin endpoint and I will keep you posted on what I find. I was looking through the admin endpoint handler and it looks like you can add a URL parameter `grace` that gives you the option to provide a grace period for shutdown. For example, if you issue a POST request to linkerd with the URL `admin/shutdown/grace=10.seconds`.  Is there a way you could test that out?

**Comment by jacob-go** [comment]
That's awesome.
I'll test it as well today.
Thank you.

The thing is, since we're running on DC/OS, we're probably going to
shutdown/restart Linkerd via Marathon API. I'll test that as well to see
how Linkerd picks up the SIGTERM.

**Comment by jacob-go** [comment]
It seems that restarting via Marathon has an effect: 
```
I 0424 04:56:47.687 UTC THREAD1: initialized
I0424 04:58:05.672359 16581 health_checker.cpp:165] Health checking stopped
I0424 04:58:25.814765 16583 health_checker.cpp:165] Health checking stopped
W0424 04:58:25.814765 16580 logging.cpp:91] RAW: Received signal SIGTERM from process 3486 of user 0; exiting
```

That double `Health checking stopped` message appeared only once when sending commands via `/admin/shutdown`. I wonder what's actually going on. Though it does saves my requests (both short-lived and with large delays; with or without `shutdownGraceMs: 10000` in the config file), while Linkerd serves as a LB, not as a local proxy on an agent.

**Comment by jacob-go** [comment]
So I've further tested. It seems that if running in DC/OS from a `buoyantio/linkerd:1.3.0` image, one has to pass a `command` to be run from `/bin/sh -c` after the container is up. This is part of the Marathon App Definition. When you pass this command to Linkerd as a parameter in Marathon's JSON, you have to specify the location of `bundle-exec` (`/io.buoyant/linkerd/1.3.0/bundle-exec`) and a path to the configuration file. This is the only way, currently, it can run.

This JSON objects looks like, for instance:
```
{
    "id": "basic-0", 
    "cmd": "while [ true ] ; do echo 'Hello Marathon' ; sleep 5 ; done",
    "cpus": 0.1,
    "mem": 10.0,
    "instances": 1
}
```

It's a pure Marathon structure.

This causes the last process to be run by Linkerd container on DC/OS to be `bundle-exec` - not "pure" linkerd. Any signal I'll send will be caught by bundle-exec or even the mesos-executor, hence those lines you've asked us about:
```
logging.cpp:91] RAW: Received signal SIGTERM from process 3486 of user 0; exiting
```

I've build a private image from this docker file:
```
FROM buoyantio/linkerd:1.3.0

COPY ./config/finalConfig.yml ./config

ENTRYPOINT ["/io.buoyant/linkerd/1.3.0/bundle-exec", "-log.level=DEBUG", "./config"]
```

The configuration inside is just a basic linkerd config, without any `shutdownGraceMs` setting. *Now*, when I restart via Marathon api, I see the line:
```
I 0424 09:38:53.431 UTC THREAD32: Received SIGTERM. Shutting down ...
```

Meaning - Linkerd is the process catching the signal. I sure that when I'll build an image with a configuration like `shutdownGraceMs: 10000` I'll see the exact behaviour you've suggested here.

**Comment by jacob-go** [comment]
**Also**, If you'll use some of Marathon's features you can still use the official buoyantIO image. 

1. use `Fetch` to download the configuration file from a given bucket/repo, for instance:
```
  "fetch": [
    {
      "uri": "https://s3.eu-central-1.amazonaws.com/<BUCKET>/config.yml",
      "extract": false,
      "executable": false,
      "cache": false
    },
```

2. Inside the `container` object of the app definition there's a `volumes` field, there you can mount the config file from MESOS_SANDBOX :
```
  "container": {
    "type": "DOCKER",
    "volumes": [
      {
        "containerPath": "/io.buoyant/linkerd/1.3.0/config.yml",
        "hostPath": "./config.yml",
        "mode": "RW"
      }
    ],
    "docker": {
      "image": "buoyantio/linkerd:1.3.0",
      "network": "HOST",
      "portMappings": [],
      "privileged": true,
      "parameters": [],
      "forcePullImage": false
    }
  },
```

3. Using the `args` object, you can pass the config file path to Linkerd's executable, as needed in any image:
```
  "args": [
    "./config.yml"
  ],
```

4. Make sure `cmd` object is empty or even not written in the App Definition. 

That way, Linkerd is the process catching signals, and the config file controls it fully.

**Comment by siggy** [comment]
@jacob-go I agree that using fetch is a more elegant solution. We echo JSON in our examples because we don't want to be responsible for hosting linkerd config files for users. For your own production setup, switching to fetch is a great option.

**Comment by adleong** [comment]
I'm glad we got to the bottom of this!  I propose that we use #1910 to track fixing the documentation but that we _also_ keep this issue open to track changing the behavior of the shutdown endpoint so that it respects the default `shutdownGraceMs` setting from the config.

This should hopefully be straightforward to implement by overriding `defaultCloseGracePeriod` in linkerd/main and namerd/main.

**Comment by adleong** [comment]
Fixed by #1951
